Dropwizard testing module for JUnit 5

AnDyXX opened the pull request:
This PR is proposition of Dropwizard testing module with use of JUnit 5 (Issue [#1944](https://github.com/dropwizard/dropwizard/issues/1944)). Contains following:
- code of original test rules was changed to support executing them from JUnit 5 extension,
- JUnit 5 extension class that executes classes from point above,
- code of original tests modified to use JUnit 5.

Example of use:

```

@ExtendWith(DropwizardExtensionsSupport.class)
public class EpisodesResourceTest {

    public final ResourceExtension resources = ResourceExtension.builder()
            .addResource(new UsersResource(usersService))
            .build();
			
```

nickbabcock commented:
Hey @AnDyXX, thanks for kicking this off! Much appreciated.

> I know that there is a lot of copied code from testing module that could be extracted to common testing library.

Is it possible to achieve backwards compatibility for the current junit4 users with a common core module? If not, we'd have to wait for something like a major version bump to introduce this change.

I'm guessing it's not possible to create the `dropwizard-testing-junit5` as a tacked on module to `dropwizard-testing` (eg. users poms would look like)

```xml
        <dependency>
            <groupId>io.dropwizard</groupId>
            <artifactId>dropwizard-testing</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>io.dropwizard</groupId>
            <artifactId>dropwizard-testing-junit5</artifactId>
            <scope>test</scope>
        </dependency>
```

As that might be the most preferable 🤔

AnDyXX commented:
> Is it possible to achieve backwards compatibility for the current junit4 users with a common core module?

Let me try if this works.

AnDyXX commented:
I changed code so JUnit4 and JUnit5 code is in the dropwizard-testing module. 
Changed junit 4 dependency from junit to junit-vintage-engine.
Common code extracted.

joschi commented:
> Changed junit 4 dependency from junit to junit-vintage-engine.

Unfortunately that makes it incompatible with JUnit and would always pull in JUnit 5 and its vintage engine.

How about making the dependencies for JUnit 4 and 5 optional?
https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html

AnDyXX commented:
Ok I made JUnit 5 optional and updated restored comments.

joschi commented:
@AnDyXX Could you please rebase this PR on current `master`?

joschi commented on dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/TestEntity.java:
```diff
@@ -34,7 +31,7 @@ long getId() {
 
     String getDescription() {
-        return description;
+        return requireNonNull(description);
```
What is that for?

Throwing a `NullPointerException` in a *getter* is very bad practice.

AnDyXX replied:
Actually this one comes from this class: https://github.com/dropwizard/dropwizard/blob/master/dropwizard-testing/src/test/java/io/dropwizard/testing/junit/TestEntity.java

joschi commented on dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/TestEntity.java:
```diff
@@ -1,12 +1,9 @@
-import javax.persistence.Table;
-import javax.validation.constraints.NotNull;
+import javax.annotation.Nullable;
+import javax.persistence.*;
```
Please don't use wildcard imports but name all imports explicitly.

joschi commented on dropwizard-testing/src/test/java/io/dropwizard/testing/junit/TestEntity.java:
```diff
@@ -1,13 +1,7 @@
-import javax.persistence.Id;
-import javax.persistence.Table;
-import javax.validation.constraints.NotNull;
+import javax.persistence.*;
```
Please don't use wildcard imports but name all imports explicitly.

AnDyXX replied:
I addressed your comments, have to look at other failing issues.

joschi commented on dropwizard-testing/src/main/java/io/dropwizard/testing/common/DAOTest.java:
```diff
@@ -0,0 +1,170 @@
+import org.hibernate.cfg.Configuration;
+import org.hibernate.context.internal.ManagedSessionContext;
+
+import java.util.*;
```
Please don't use wildcard imports but name all imports explicitly.

joschi commented on dropwizard-testing/src/main/java/io/dropwizard/testing/junit5/DropwizardExtensionsSupport.java:
```diff
@@ -0,0 +1,94 @@
+package io.dropwizard.testing.junit5;
+
+import org.junit.jupiter.api.extension.*;
```
Please don't use wildcard imports but name all imports explicitly.

joschi commented on dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/DropwizardAppExtensionTest.java:
```diff
@@ -0,0 +1,130 @@
+
+import static io.dropwizard.testing.ResourceHelpers.resourceFilePath;
+import static org.hamcrest.core.Is.is;
+import static org.hamcrest.MatcherAssert.*;
```
Please don't use wildcard imports but name all imports explicitly.

joschi commented on dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/GzipDefaultVaryBehaviourTest.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.Collections;
+
+import static io.dropwizard.testing.ResourceHelpers.resourceFilePath;
+import static javax.ws.rs.core.HttpHeaders.*;
```
Please don't use wildcard imports but name all imports explicitly.

joschi commented on dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/PersonResourceTest.java:
```diff
@@ -0,0 +1,130 @@
+import javax.ws.rs.ext.ExceptionMapper;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.*;
```
Please don't use wildcard imports but name all imports explicitly.

joschi commented on dropwizard-testing/src/test/java/io/dropwizard/testing/junit5/TestEntity.java:
```diff
@@ -0,0 +1,40 @@
+package io.dropwizard.testing.junit5;
+
+import javax.annotation.Nullable;
+import javax.persistence.*;
```
Please don't use wildcard imports but name all imports explicitly.

AnDyXX commented:
`codeclimate — 42 issues to fix` are mostly because of test for JUnit5 copied from junit. Any hints on this ones?

arteam commented:
I've manually marked warnings in Codeclimate as false positives.

AnDyXX commented:
PR code rebased.

arteam commented:
Thank you for the contribution!

krupeshf commented:
@AnDyXX / @arteam   

Do you have updated version of testing resources documentation?
https://www.dropwizard.io/1.3.9/docs/manual/testing.html#testing-resources  
Currently it is still using `JUnit4` and `ClassRule`  

Thank you :)
